File: packages/sdk-typescript/src/agent-client.ts
// SPDX-License-Identifier: Apache-2.0
import { randomUUID } from 'node:crypto'
import { existsSync } from 'node:fs'
import { join } from 'node:path'
import type { AgentCard, MessageEnvelope, ResponseEnvelope, TaskRecord } from './types.js'
import { generateKeypair, loadKeypair, saveKeypair, type Keypair } from './keys.js'
import { signRequest } from './signing.js'
import { SamvadError, ErrorCode, type ErrorCodeType } from './errors.js'

export interface AgentClientOptions {
  keysDir?: string
  agentId?: string
}

export class AgentClient {
  public card: AgentCard | null = null
  private baseUrl: string | null = null

  private constructor(
    private readonly keypair: Keypair,
    public readonly agentId: string,
  ) {}

  /** Create a client without connecting. Used when you need the public key before connecting (e.g. to trust-peer it on the target agent). */
  static async prepare(opts: AgentClientOptions = {}): Promise<AgentClient> {
    const keysDir = opts.keysDir ?? join(process.cwd(), '.samvad', 'client-keys')
    let kp: Keypair
    const keyFile = join(keysDir, 'key-current.json')
    if (existsSync(keyFile)) {
      kp = await loadKeypair(keysDir, 'key-current')
    } else {
      kp = await generateKeypair('key-current')
      await saveKeypair(kp, keysDir)
    }
    const agentId = opts.agentId ?? `agent://client.local`
    return new AgentClient(kp, agentId)
  }

  /** Connect to an already-prepared client's target agent. */
  async connect(agentUrl: string): Promise<void> {
    this.baseUrl = agentUrl
    const cardRes = await fetch(`${agentUrl}/.well-known/agent.json`)
    if (!cardRes.ok) throw new Error(`Failed to fetch agent card from ${agentUrl}: ${cardRes.status}`)
    this.card = await cardRes.json() as AgentCard
  }

  /** Convenience: prepare + connect in one call. Used when you don't need the public key in advance. */
  static async from(agentUrl: string, opts: AgentClientOptions = {}): Promise<AgentClient> {
    const client = await AgentClient.prepare(opts)
    await client.connect(agentUrl)
    return client
  }

  get publicKey(): Uint8Array {
    return this.keypair.publicKey
  }

  private ensureConnected(): void {
    if (!this.baseUrl || !this.card) {
      throw new Error('AgentClient is not connected. Call connect(agentUrl) first.')
    }
  }

  private buildBody(
    skill: string,
    mode: 'sync' | 'async' | 'stream',
    payload: Record<string, unknown>,
  ): MessageEnvelope {
    this.ensureConnected()
    return {
      from: this.agentId,
      to: this.card!.id,
      skill, mode,
      nonce: randomUUID(),
      timestamp: new Date().toISOString(),
      traceId: randomUUID(),
      spanId: randomUUID(),
      payload,
    }
  }

  /** Sign and send a POST request with RFC 9421 signature headers. */
  private async signedPost(url: string, body: object): Promise<Response> {
    const bodyStr = JSON.stringify(body)
    const bodyBytes = Buffer.from(bodyStr)
    const { pathname } = new URL(url)
    const sigHeaders = await signRequest('POST', pathname, bodyBytes, this.keypair)
    return fetch(url, {
      method: 'POST',
      headers: { 'content-type': 'application/json', ...sigHeaders },
      body: bodyStr,
    })
  }

  async call(skill: string, payload: Record<string, unknown>): Promise<Record<string, unknown>> {
    this.ensureConnected()
    const body = this.buildBody(skill, 'sync', payload)
    const res = await this.signedPost(`${this.baseUrl}/agent/message`, body)
    const respBody = await res.json() as ResponseEnvelope
    if (respBody.status === 'error') {
      throw new SamvadError(respBody.error!.code as ErrorCodeType, respBody.error!.message)
    }
    return respBody.result!
  }

  async task(skill: string, payload: Record<string, unknown>, callbackUrl?: string): Promise<string> {
    this.ensureConnected()
    const body: Record<string, unknown> = { ...this.buildBody(skill, 'async', payload) }
    if (callbackUrl) body.callbackUrl = callbackUrl
    const res = await this.signedPost(`${this.baseUrl}/agent/task`, body)
    const result = await res.json() as { taskId: string; status: string }
    return result.taskId
  }

  async taskAndPoll(
    skill: string,
    payload: Record<string, unknown>,
    opts: { intervalMs?: number; timeoutMs?: number } = {},
  ): Promise<Record<string, unknown>> {
    this.ensureConnected()
    const taskId = await this.task(skill, payload)
    const interval = opts.intervalMs ?? 100
    const timeout = opts.timeoutMs ?? 30_000
    const deadline = Date.now() + timeout

    while (Date.now() < deadline) {
      await new Promise(r => setTimeout(r, interval))
      const res = await fetch(`${this.baseUrl}/agent/task/${taskId}`)
      const task = await res.json() as TaskRecord
      if (task.status === 'done') return task.result!
      if (task.status === 'failed') {
        throw new SamvadError(task.error!.code as ErrorCodeType, task.error!.message)
      }
    }
    throw new SamvadError(ErrorCode.AGENT_UNAVAILABLE, `Task ${taskId} timed out after ${timeout}ms`)
  }

  async *stream(skill: string, payload: Record<string, unknown>): AsyncGenerator<unknown> {
    this.ensureConnected()
    const body = this.buildBody(skill, 'stream', payload)
    const url = `${this.baseUrl}/agent/stream`
    const bodyStr = JSON.stringify(body)
    const bodyBytes = Buffer.from(bodyStr)
    const { pathname } = new URL(url)
    const sigHeaders = await signRequest('POST', pathname, bodyBytes, this.keypair)
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'content-type': 'application/json', ...sigHeaders },
      body: bodyStr,
    })
    if (!res.body) throw new Error('No response body for stream')
    const reader = res.body.getReader()
    const decoder = new TextDecoder()
    let buffer = ''

    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n\n')
      buffer = lines.pop() ?? ''
      for (const line of lines) {
        if (line.startsWith('data: ')) {
          const data = JSON.parse(line.slice(6))
          if (data.done) {
            if (data.error) throw new SamvadError(data.error.code as ErrorCodeType, data.error.message)
            return
          }
          yield data.chunk
        }
      }
    }
  }
}

File: packages/sdk-python/src/samvad/agent_client.py
# SPDX-License-Identifier: Apache-2.0
from __future__ import annotations

import asyncio
import json
import time
import uuid
from pathlib import Path
from typing import Any, AsyncIterator

import httpx

from .keys import Keypair, load_or_generate_keypair
from .signing import canonical_json, content_digest, sign_request
from .types import AgentCard, MessageEnvelope


class AgentClient:
    def __init__(self, *, keypair: Keypair, card: AgentCard | None = None) -> None:
        self._kp = keypair
        self._card = card

    @classmethod
    def prepare(
        cls,
        *,
        keys_dir: str = ".samvad/client-keys/",
        kid: str = "client",
    ) -> "AgentClient":
        kp = load_or_generate_keypair(Path(keys_dir), kid)
        return cls(keypair=kp)

    @classmethod
    async def from_url(
        cls,
        url: str,
        *,
        keys_dir: str = ".samvad/client-keys/",
        kid: str = "client",
    ) -> "AgentClient":
        client = cls.prepare(keys_dir=keys_dir, kid=kid)
        await client.connect(url)
        return client

    async def connect(self, url: str) -> None:
        card_url = url.rstrip("/") + "/.well-known/agent.json"
        async with httpx.AsyncClient() as h:
            resp = await h.get(card_url, timeout=10)
            resp.raise_for_status()
            self._card = AgentCard.model_validate(resp.json())

    @property
    def public_key_b64(self) -> str:
        return self._kp.public_key_b64

    @property
    def kid(self) -> str:
        return self._kp.kid

    def _build_envelope(
        self,
        skill: str,
        payload: dict[str, Any],
        mode: str = "sync",
        callback_url: str | None = None,
    ) -> dict[str, Any]:
        agent_id = f"agent://client-{self._kp.kid}"
        target_id = self._card.id if self._card else "agent://unknown"
        env: dict[str, Any] = {
            "from": agent_id,
            "to": target_id,
            "skill": skill,
            "mode": mode,
            "nonce": str(uuid.uuid4()),
            "timestamp": _iso_now(),
            "traceId": str(uuid.uuid4()),
            "spanId": str(uuid.uuid4()),
            "payload": payload,
        }
        if callback_url:
            env["callbackUrl"] = callback_url
        return env

    def _sign_envelope(
        self, envelope: dict[str, Any], method: str, path: str
    ) -> tuple[bytes, dict[str, str]]:
        raw_body = canonical_json(envelope).encode("utf-8")
        digest = content_digest(raw_body)
        base_headers = {
            "content-type": "application/json",
            "content-digest": digest,
        }
        sig = sign_request(method, path, base_headers, self._kp.private_key_b64, self._kp.kid)
        return raw_body, {**base_headers, **sig}

    async def call(self, skill: str, payload: dict[str, Any]) -> dict[str, Any]:
        if self._card is None:
            raise RuntimeError("call connect() or from_url() first")
        base_url = self._card.url.rstrip("/")
        env = self._build_envelope(skill, payload, mode="sync")
        raw_body, headers = self._sign_envelope(env, "POST", "/agent/message")
        async with httpx.AsyncClient() as h:
            resp = await h.post(
                f"{base_url}/agent/message",
                content=raw_body,
                headers=headers,
                timeout=30,
            )
            resp.raise_for_status()
            return resp.json()

    async def task(
        self, skill: str, payload: dict[str, Any], *, callback_url: str | None = None
    ) -> str:
        if self._card is None:
            raise RuntimeError("call connect() or from_url() first")
        base_url = self._card.url.rstrip("/")
        env = self._build_envelope(skill, payload, mode="async", callback_url=callback_url)
        raw_body, headers = self._sign_envelope(env, "POST", "/agent/task")
        async with httpx.AsyncClient() as h:
            resp = await h.post(f"{base_url}/agent/task", content=raw_body, headers=headers, timeout=30)
            resp.raise_for_status()
            return resp.json()["taskId"]

    async def task_and_poll(
        self,
        skill: str,
        payload: dict[str, Any],
        *,
        interval: float = 0.5,
        timeout: float = 60.0,
    ) -> dict[str, Any]:
        if self._card is None:
            raise RuntimeError("call connect() or from_url() first")
        task_id = await self.task(skill, payload)
        base_url = self._card.url.rstrip("/")
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            await asyncio.sleep(interval)
            async with httpx.AsyncClient() as h:
                resp = await h.get(f"{base_url}/agent/task/{task_id}", timeout=10)
                resp.raise_for_status()
                record = resp.json()
                if record.get("status") in ("done", "failed"):
                    return record
        raise TimeoutError(f"task {task_id} did not complete within {timeout}s")

    async def stream(
        self, skill: str, payload: dict[str, Any]
    ) -> AsyncIterator[dict[str, Any]]:
        if self._card is None:
            raise RuntimeError("call connect() or from_url() first")
        base_url = self._card.url.rstrip("/")
        env = self._build_envelope(skill, payload, mode="stream")
        raw_body, headers = self._sign_envelope(env, "POST", "/agent/stream")
        async with httpx.AsyncClient() as h:
            async with h.stream("POST", f"{base_url}/agent/stream",
                                content=raw_body, headers=headers, timeout=60) as resp:
                resp.raise_for_status()
                async for line in resp.aiter_lines():
                    if line.startswith("data: "):
                        yield json.loads(line[6:])


def _iso_now() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
